### PR TITLE
Allow keywords to use Web Elements in addition to Locator.

### PR DESCRIPTION
**Pull request by emanlove**

### CODE REVIEW

File: src/Selenium2Library/__init__.py [hunk 1]
```diff
@@ -69,6 +78,17 @@ class Selenium2Library(
     This can be fixed by changing the locator to:
     | Click Link    default=page?a=b
 
+    Using 'webelement'
+    ------------------
+    Starting with version 1.6 of the Selenium2Library, one can pass an argument
```
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py]: This should be 1.7 instead of 1.6.

File: src/Selenium2Library/__init__.py [hunk 2]
```diff
@@ -81,7 +101,12 @@ class Selenium2Library(
     | css        | Table Should Contain `|` css=table.my_class `|` text               | Matches by @id or @name attribute |
     | xpath      | Table Should Contain `|` xpath=//table/[@name="my_table"] `|` text | Matches by @id or @name attribute |
 
-    = Timeouts =
+    *Stale and Detached Elements*
```
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py]: What needs to go here? We should probably flesh out these docs before merge.

File: src/Selenium2Library/keywords/_element.py [hunk 3]
```diff
@@ -588,12 +606,17 @@ def xpath_should_match_x_times(self, xpath, expected_xpath_count, message='', lo
...
+            elements = self._element_finder.find(browser, locator, tag)
+            if required and len(elements) == 0:
+                raise ValueError("Element locator '" + locator + "' did not match any elements.")
+            if first_only:
+                if len(elements) == 0: return None
+                return elements[0]
+        elif isinstance(locator, WebElement):
+            elements = locator
+        # do some other stuff here like deal with list of webelements
+        # ... or raise locator/element specific error if required
```
Comment (just-be-dev) [inline src/Selenium2Library/keywords/_element.py:669]: Is there anything specific that goes here, or is this just a future note?

**Comment by just-be-dev** [comment]
When you get a chance, could you rebase? I fixed the Jenkins build failing.

File: src/Selenium2Library/keywords/_element.py [hunk 4]
```diff
@@ -1,14 +1,32 @@
...
+        return isinstance(s, str)
+
 class _ElementKeywords(KeywordGroup):
 
     def __init__(self):
         self._element_finder = ElementFinder()
 
+    # Public, get element(s)
+
+    def get_webelements(self, locator):
```
Comment (just-be-dev) [inline src/Selenium2Library/keywords/_element.py:24]: Instead of `Get WebElements`, how would you feel about just `Get Elements` . Also, would it be useful to have a `Get Element` that only gets a single element, or would that be unnecessary/confusing?

File: src/Selenium2Library/__init__.py (hunk 2 above)
Comment (emanlove) [inline src/Selenium2Library/__init__.py]: Thought I had pulled that out. It was a placeholder for talking about stale elements [as recommended above](https://github.com/rtomac/robotframework-selenium2library/pull/356#commitcomment-8126221) by @ombre42. Since it was taking too long to compose the doc I was going to add it later...

File: src/Selenium2Library/keywords/_element.py (hunk 3 above)
Comment (emanlove) [inline src/Selenium2Library/keywords/_element.py:669]: Just a future note...

File: src/Selenium2Library/keywords/_element.py (hunk 4 above)
Comment (emanlove) [inline src/Selenium2Library/keywords/_element.py:24]: @ombre42 and I discussed this over on #315. 'Get Elements' was my original name but changed it because it conflicted with the XML Library's keyword of the same name. So we settled on Get WebElements.  I also had a 'Get WebElement' (singular) keyword in there but removed it for some reason. I was doing some recent testing with this and saw a possible need for getting a single web element.

File: src/Selenium2Library/__init__.py (hunk 1 above)
Comment (emanlove) [inline src/Selenium2Library/__init__.py]: You are correct. How time flies!

File: src/Selenium2Library/keywords/_element.py (hunk 4 above)
Comment (just-be-dev) [inline src/Selenium2Library/keywords/_element.py:24]: How about `Find Element` and `Find Elements` then? It's not necessarily as clear, but that matches up with the selenium API. (I'm not really necessarily opposed to `Get WebElements` though).

File: src/Selenium2Library/__init__.py (hunk 2 above)
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py]: I'm okay with that, but we need to make a new issue for it so we don't forget.

**Comment by just-be-dev** [comment]
After you rebase, I'm ready to merge this in.

File: src/Selenium2Library/__init__.py [hunk 5]
```diff
@@ -34,13 +34,22 @@ class Selenium2Library(
...
+    **--- Note important change starting with Version 1.7.0 release ---**
+    *Locating or getting elements*
+    *Locating or specifying elements*
 
     All keywords in Selenium2Library that need to find an element on the page
-    take an argument, `locator`. By default, when a locator value is provided,
-    it is matched against the key attributes of the particular element type.
-    For example, `id` and `name` are key attributes to all elements, and
-    locating elements is easy using just the `id` as a `locator`. For example::
+    take an argument, either a `locator` or now a 'webelement'. 'locator'
```
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py]: webelement and locator should be wrapped with backticks, not apostrophes.

File: src/Selenium2Library/__init__.py [hunk 6]
```diff
@@ -34,13 +34,22 @@ class Selenium2Library(
     imported into your Robot test suite (see `importing` section), and the 
     `Open Browser` keyword must be used to open a browser to the desired location.
 
-    = Locating elements =
+    **--- Note important change starting with Version 1.7.0 release ---**
+    *Locating or getting elements*
+    *Locating or specifying elements*
```
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py]: Are these two lines accidental duplication, or are they on purpose? It seems like you either want 38 or 39 but not both. Also, why did you remove the = sign in favor of asterisks?

File: src/Selenium2Library/__init__.py [hunk 7]
```diff
@@ -34,13 +34,22 @@ class Selenium2Library(
...
-    take an argument, `locator`. By default, when a locator value is provided,
-    it is matched against the key attributes of the particular element type.
-    For example, `id` and `name` are key attributes to all elements, and
-    locating elements is easy using just the `id` as a `locator`. For example::
+    take an argument, either a `locator` or now a 'webelement'. 'locator'
+    is a string that describes how to locate an element using a syntax
+    specifying different location strategies. 'webelement' is a variable that
+    holds a WebElement instance, which is a representation of the element.
+
+    Using 'locator'
```
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py:45]: backticks

File: src/Selenium2Library/__init__.py [hunk 8]
```diff
@@ -81,7 +101,12 @@ class Selenium2Library(
     | css        | Table Should Contain `|` css=table.my_class `|` text               | Matches by @id or @name attribute |
     | xpath      | Table Should Contain `|` xpath=//table/[@name="my_table"] `|` text | Matches by @id or @name attribute |
 
-    = Timeouts =
+    *Stale and Detached Elements*
+
+    To be written... a talk about stale/detached elements, the dreaded
+    StaleElementReferenceException and how to deal with these situations.
+
+    *Timeouts*
```
Comment (just-be-dev) [inline src/Selenium2Library/__init__.py]: Again, why \* instead of = ?

File: src/Selenium2Library/keywords/_element.py [hunk 9]
```diff
@@ -1,14 +1,32 @@
...
+        """Returns list of WebElement objects matching locator.
+
+        See `introduction` for details about locating elements.
+        """
+        return self._element_find(locator, False, True)
+
     # Public, element lookups
```
Comment (just-be-dev) [inline src/Selenium2Library/keywords/_element.py:32]: I still think `get_webelement` would be useful here.

File: src/Selenium2Library/__init__.py (hunk 6 above)
Comment (emanlove) [inline src/Selenium2Library/__init__.py]: Already corrected. It was a left over artifact of my indecision at the time. Not sure about the asterisks; can't remember.

File: src/Selenium2Library/keywords/_element.py (hunk 9 above)
Comment (emanlove) [inline src/Selenium2Library/keywords/_element.py:32]: Agreed.

**Comment by just-be-dev** [comment]
Overall this looks good. I'm good for you to merge whenever you're ready.

**Comment by zephraph** [comment]
@emanlove, would you be okay if I went ahead and merged this? The documentation nitpicks will be fixed on the next update anyway, and I can add a PR to do the ```Get WebElement``` Keyword.

**Comment by emanlove** [comment]
On 05/11/2015 10:49 AM, Zephraph wrote:

> @emanlove https://github.com/emanlove, would you be okay if I went 
> ahead and merged this? The documentation nitpicks will be fixed on the 
> next update anyway, and I can add a PR to do the |Get WebElement| Keyword.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/rtomac/robotframework-selenium2library/pull/356#issuecomment-100932596.

Yes, please merge in.

**Comment by just-be-dev** [comment]
It seems you need to rebase before I can do so.

**Comment by emanlove** [comment]
Merged with the minor documentation corrections added in.

**Comment by just-be-dev** [comment]
And that's a wrap! Great work! I want to add some integration tests before the 1.7 release, but we're super close!
